feat(SPSTRAT-482): add pre-push for marketplacesvm

JAVGan opened the pull request:
This commit allows the `marketplacesvm-push-disk-images` task to set the parameter `nochannel` on the wrapper to use the `pre-push` mode.

The `pre-push` mode will cause the VM images to be uploaded to the respective clouds but it won't fully publish on them. This is ideal to perform the "staging" tests before the final push.

It also fixes a naming issue with the task
`collect-marketplacesvm-secret`.

Refers to [SPSTRAT-482](https://issues.redhat.com//browse/SPSTRAT-482)

## Describe your changes

## Relevant Jira

## Checklist before requesting a review
- [x] I have marked as draft or added `do not merge` label if there's a dependency PR
  - If you want reviews on your draft PR, you can add reviewers or add the `release-service-maintainers` handle if you are unsure who to tag
- [x] My commit message includes `Signed-off-by: My name <email>`
- [x] I have bumped the task/pipeline version string and updated changelog in the relevant README
- [x] I read CONTRIBUTING.MD and [commit formatting](CONTRIBUTING.md#commit-message-formatting-and-standards)

JAVGan committed:
feat(SPSTRAT-482): add pre-push for marketplacesvm

This commit allows the `marketplacesvm-push-disk-images` task to set the
parameter `nochannel` on the wrapper to use the `pre-push` mode.

The `pre-push` mode will cause the VM images to be uploaded to the
respective clouds but it won't fully publish on them. This is ideal to
perform the "staging" tests before the final push.

It also fixes a naming issue with the task
`collect-marketplacesvm-secret`.

Refers to SPSTRAT-482

Signed-off-by: Jonathan Gangi <jgangi@redhat.com>

mmalina commented on pipelines/push-disk-images-to-marketplaces/README.md:
```diff
@@ -18,3 +18,6 @@ Tekton Pipeline to push disk images to various cloud marketplaces
 | taskGitRevision                 | The revision in the taskGitUrl repo to be used                                                         | No       | -                                                         |
+
+## Changes in 0.1.1
+* Allow `marketplacesvm-push-disk-images` task o run on pre-push mode.
```
```suggestion
* Allow `marketplacesvm-push-disk-images` task to run in pre-push mode.
```

JAVGan replied:
Fixed

mmalina commented on pipelines/push-disk-images-to-marketplaces/push-disk-images-to-marketplaces.yaml:
Would it make sense to also rename `push-disk-images-marketplaces` to `marketplacesvm-push-disk-images` to match the task name?

JAVGan replied:
Yes, I can do that! In this case the suggestion is to rename the pipeline, right?

mmalina replied:
No, the suggestion was to rename the task name in the pipeline definition to match the actual task name:
```
tasks:
  - name: push-disk-images-marketplaces
```
to 
```
tasks:
  - name: marketplacesvm-push-disk-images
```

JAVGan replied:
Oh, ok!

Fixed that

mmalina commented on pipelines/push-disk-images-to-marketplaces/push-disk-images-to-marketplaces.yaml:
```diff
@@ -227,6 +227,8 @@ spec:
         - name: cloudMarketplacesSecret
           value: "$(tasks.collect-marketplacesvm-secret.results.cloudMarketplacesSecret)"
+        - name: prePush
+          value: "$(tasks.collect-marketplacesvm-secret.results.prePush)"
```
Would it make sense to rename `collect-marketplacesvm-secret` to `collect-marketplacesvm-params` since it's now collecting more than just the secret?

JAVGan replied:
Sure, I can do that!

JAVGan replied:
Renamed

mmalina commented on tasks/collect-marketplacesvm-secret/collect-marketplacesvm-secret.yaml:
```diff
@@ -42,3 +45,5 @@ spec:
         fi
 
         jq -j '.mapping.cloudMarketplacesSecret' "$DATA_FILE" | tee "$(results.cloudMarketplacesSecret.path)"
+        jq -jr '.mapping | if has("cloudMarketplacesPrePush") then .cloudMarketplacesPrePush else "false" end' \
```
Can we not do this? Seems simpler. Also, `-j` and `-r` don't make sense together - `-j` is enough.
```suggestion
        jq -j '.mapping.cloudMarketplacesPrePush // false' \
```

JAVGan replied:
I've tried that but it seems it fails when the key is absent...

mmalina replied:
Why would it fail? We use that in many places. For example:
```
$ jq -j '.mapping.cloudMarketplacesPrePush // false' <<< '{}'
false
```

JAVGan replied:
Nevermind.. It's working now, I'm patching it

JAVGan replied:
Fixed! It's working 😄

mmalina commented on tasks/marketplacesvm-push-disk-images/README.md:
```diff
@@ -10,4 +10,8 @@ It currently supports images in `raw` and `vhd` formats for `AWS` and `Azure` re
 | ----------------------- | -------------------------------------------------------------------------------------- | -------- | --------------- |
 | snapshotPath            | Path to the JSON string of the mapped snapshot spec in the data workspace.             | No       | -               |
 | cloudMarketplacesSecret | Env specific secret containing the marketplaces credentials.                           | No       | -               |
+| prePush                 | Whether perform a pre-push (true) or not (false). When true it will not publish PROD.  | No       | false           |
```
It has default, so it is optional, right?

JAVGan replied:
Yes, it is! My mistake here, I'll adjust it

JAVGan replied:
Adjusted

mmalina commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -4,7 +4,7 @@ kind: Task
   name: marketplacesvm-push-disk-images
   labels:
-    app.kubernetes.io/version: "0.1.0"
+    app.kubernetes.io/version: "0.1.1"
```
It's not that important, but I think it should be 0.2.0 (also in README) - it's a new feature, not a bug fix.

JAVGan replied:
Changed

mmalina commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -140,4 +144,9 @@ spec:
+        if [[ "${pre_push,,}" == "true" ]]; then
+          add_args="--nochannel"
+        fi
+        marketplacesvm_push_wrapper --debug $add_args --source "${BASE_DIR}" --starmap-file "$STARMAP_MAPPING_FILE"
```
I'm surprised shellcheck does not complain about `$add_args` :) So I guess it's fine. Otherwise you could use an array for this, e.g. 
```
EXTRA_ARGS=()
if prepush; then
  EXTRA_ARGS=(--nochannel)
marketplacesvm_push_wrapper ... "${EXTRA_ARGS[@]}"` ...
```

But ok to leave as is, up to you.

JAVGan replied:
Thanks! I've added this, it's better

mmalina commented on tasks/marketplacesvm-push-disk-images/marketplacesvm-push-disk-images.yaml:
```diff
@@ -140,4 +144,9 @@ spec:
-        marketplacesvm_push_wrapper --debug --source "${BASE_DIR}" --starmap-file "$STARMAP_MAPPING_FILE"
+        pre_push="$(params.prePush)"
+        add_args=""
+        if [[ "${pre_push,,}" == "true" ]]; then
```
What are the `,,` for?

JAVGan replied:
To put the value in lowercase so it can do the string comparison without being case sensitive

mmalina replied:
I see. I'd say it's probably not needed. We have a number of options like this and we never allow uppercase. But this brings me to another thing: The new option should be added to the schema file. That will also ensure that it's the right format - if you specify it as boolean, then reading it with `jq` should always return the boolean "true" or "false".

JAVGan replied:
Adjusted!

JAVGan commented:
@mmalina PTAL again.

I believe I've adjusted everything you asked, please let me know if there's something else or if I did miss anything

mmalina commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images-pre-push.yaml:
```diff
@@ -0,0 +1,152 @@
+  name: test-marketplacesvm-push-disk-images
+spec:
+  description: |
+    Run the marketplacesvm-push-disk-images task with valid data to ensure it succeeds
```
This should mention that it's the pre-push scenario.

mmalina reviewed (commented):
Everything looks good, except one small thing - a test description. Once that's fixed, it will be ready to go.

JAVGan commented:
@mmalina I've adjusted the test but I also rebased it and noticed some files have changed location, hopefully it's ok. Please let me know if there's something wrong with it

mmalina commented:
> @mmalina I've adjusted the test but I also rebased it and noticed some files have changed location, hopefully it's ok. Please let me know if there's something wrong with it

You mean because of this, right? https://github.com/konflux-ci/release-service-catalog/pull/749

Yeah, that should be fine.

JAVGan commented:
@mmalina feel free to merge it whenever it's ok for you

johnbieren commented on tasks/collect-marketplacesvm-params/tests/test-collect-marketplacesvm-params-pre-push.yaml:
```diff
@@ -0,0 +1,79 @@
+              #!/usr/bin/env sh
+              set -eux
+
+              cat > $(workspaces.data.path)/data.json << EOF
```
@mmalina how is checkton not complaining for this?

mmalina replied:
Good question... @chmeliik any idea? In the workflow log I can see:
```
Files to check:
...
C079	tasks/collect-marketplacesvm-secret/tests/test-collect-marketplacesvm-secret.yaml	tasks/collect-marketplacesvm-params/tests/test-collect-marketplacesvm-params-pre-push.yaml
```

JAVGan replied:
What is the issue here?

johnbieren replied:
We expect checkton to report that you don't have `"` around `$(workspaces.data.path)/data.json`. We have been fixing checkton in all of our tasks and pretty much every task complained about it. But checkton is passing here, which we find strange

JAVGan replied:
Gotcha! Should I patch it or leave this for you to investigate more on checkton?

johnbieren replied:
Maybe it is due to the use of `sh` instead of `bash`? checkton passes so it is okay for you to leave it. We don't want to start mandating things in addition to what checkton tells you to do

mmalina replied:
> Maybe it is due to the use of sh instead of bash?

I don't think so. If I copy and paste the first few lines at https://www.shellcheck.net it will complain.

chmeliik replied:
> Good question... @chmeliik any idea?

Yeah, what happened here is this:

```
  C079	tasks/collect-marketplacesvm-secret/tests/test-collect-marketplacesvm-secret.yaml	tasks/collect-marketplacesvm-params/tests/test-collect-marketplacesvm-params-pre-push.yaml
  R082	tasks/collect-marketplacesvm-secret/tests/test-collect-marketplacesvm-secret.yaml	tasks/collect-marketplacesvm-params/tests/test-collect-marketplacesvm-params.yaml
```

The base branch had a `tasks/collect-marketplacesvm-secret/tests/test-collect-marketplacesvm-secret.yaml` file. This PR has two files, one is an 79% Copy and the other is an 82% Rename of the file from the base branch.

Checkton compares shellcheck warnings with those that already exist in the base branch and doesn't report them. This include handling for copies and renames (warnings in a copied/renamed file won't get reported if the base file already had them)

You can tweak the behavior with some params if you want https://github.com/chmeliik/checkton?tab=readme-ov-file#find-renames

johnbieren replied:
Okay, @JAVGan can you add the quotes

JAVGan replied:
Sure, adding them

JAVGan replied:
I've added them, PTAL again.

I've also rebased the PR

mmalina replied:
@chmeliik thanks for the explanation. But doesn't that mean that this should have been reported when the file in the base branch was added? See here: https://github.com/konflux-ci/release-service-catalog/blob/a7ee201db4bee2044c8b1f0e325302ad8df1e896/tasks/collect-marketplacesvm-secret/tests/test-collect-marketplacesvm-secret.yaml#L27

It was added here: https://github.com/konflux-ci/release-service-catalog/pull/719

That PR has the line `cat > $(workspaces.data.path)/data.json << EOF`.

From the checkton run there:
```
C069	tasks/collect-registry-token-secret/tests/test-collect-registry-token-secret.yaml	tasks/collect-marketplacesvm-secret/tests/test-collect-marketplacesvm-secret.yaml
```

So once again, it treated it as a copy of existing `tasks/collect-registry-token-secret/tests/test-collect-registry-token-secret.yaml`. And that one had the same line. It was added in September: https://github.com/konflux-ci/release-service-catalog/pull/523 Checkton succeded there, but there are no logs - probably too old.

So that's interesting.

This wouldn't be an issue if everything our repo was fixed for checkton. We're working on that here and it's almost done: https://issues.redhat.com/browse/RELEASE-1032

@johnbieren maybe we should add a jira to run checkton on everything once all the other stories are done and see if there are any leftovers - there likely will be because of this.

chmeliik replied:
Yeah, if all the test files are roughly copies of each other, the default behavior will end up hiding a lot of warnings. You may want to disable the handling for copies or to disable differential mode altogether once you've fixed all the warnings

mmalina replied:
If we didn't work on fixing everything, I would say we should set find-renames and find-copies to false. (BTW, @johnbieren you were worried that if you moved all the files before we finish fixing this, you would have all the checkton warnings - as we can see here, you wouldn't ;) )

> You may want to disable the handling for copies or to disable differential mode altogether once you've fixed all the warnings

Good point. I am not sure we should do it. Maybe... That would be similar to how tests in konflux-release-data are run. On the other hand in this repo we only run tekton task unit tests for tasks that are being changed. So maybe it makes sense to leave checkton similar to that.

johnbieren replied:
> @johnbieren maybe we should add a jira to run checkton on everything once all the other stories are done and see if there are any leftovers - there likely will be because of this.

+1

I am fine with us doing a one off to fix lingering things. None of the checkton issues were that hard to fix, and there are only a few unique things to do so we can just copy what we did in other tasks. In other words, that ticket should be straight forward.

Going forward, I'd like checkton to run on any file in the diff. So if you update line 8 of file x/y/z, z should fail checkton even if the violating line is in line 20 of file z or if z is similar to file x/a/b. Is that a possible config setting @chmeliik ?

chmeliik replied:
If I understand correctly, you want to disable finding renames and copies https://github.com/chmeliik/checkton?tab=readme-ov-file#find-renames

johnbieren replied:
https://github.com/konflux-ci/release-service-catalog/pull/766

johnbieren commented on tasks/marketplacesvm-push-disk-images/README.md:
```diff
@@ -10,4 +10,8 @@ It currently supports images in `raw` and `vhd` formats for `AWS` and `Azure` re
 | concurrentLimit         | The maximum number of images to be pulled at once.                                     | Yes      | 3               |
+
+## Changes in 0.2.0
+* Allow `marketplacesvm-push-disk-images` task o run on pre-push mode.
```
`s/ o / to /`

JAVGan replied:
Nice catch! Fixed

johnbieren commented on tasks/marketplacesvm-push-disk-images/tests/test-marketplacesvm-push-disk-images-pre-push.yaml:
```diff
@@ -0,0 +1,152 @@
+apiVersion: tekton.dev/v1
+kind: Pipeline
+metadata:
+  name: test-marketplacesvm-push-disk-images
```
This needs to match the filename

JAVGan replied:
Fixed, thanks!

johnbieren commented:
/ok-to-test

johnbieren reviewed (approved):
There is no e2e test for this pipeline. Merging
